File: src/modules/trading/handlers.js
/**
 * @file Command handler implementations for the trading module.
 * Each handler receives (ctx, db) — the grammY context and KV store.
 * Currently only VN stocks are supported. Crypto/gold/convert coming later.
 */

import { formatStock, formatVND } from "./format.js";
import {
  addAsset,
  addCurrency,
  deductAsset,
  deductCurrency,
  getPortfolio,
  savePortfolio,
} from "./portfolio.js";
import { getStockPrice } from "./prices.js";
import { comingSoonMessage, resolveSymbol } from "./symbols.js";

function uid(ctx) {
  return ctx.from?.id;
}

function parseArgs(ctx) {
  return (ctx.match || "").trim().split(/\s+/).filter(Boolean);
}

function usageReply(ctx, usage) {
  return ctx.reply(`Usage: ${usage}`);
}

/** /trade_topup <amount> — add VND to account */
export async function handleTopup(ctx, db) {
  const args = parseArgs(ctx);
  if (args.length < 1)
    return usageReply(ctx, "/trade_topup <amount>\nExample: /trade_topup 5000000");
  const amount = Number(args[0]);
  if (!Number.isFinite(amount) || amount <= 0)
    return ctx.reply("Amount must be a positive number.");

  const p = await getPortfolio(db, uid(ctx));
  addCurrency(p, "VND", amount);
  p.meta.invested += amount;
  await savePortfolio(db, uid(ctx), p);
  await ctx.reply(`Topped up ${formatVND(amount)}.\nBalance: ${formatVND(p.currency.VND)}`);
}

/** /trade_buy <amount> <symbol> — buy VN stock at market price */
export async function handleBuy(ctx, db) {
  const args = parseArgs(ctx);
  if (args.length < 2)
    return usageReply(ctx, "/trade_buy <qty> <TICKER>\nExample: /trade_buy 100 TCB");
  const amount = Number(args[0]);
  if (!Number.isFinite(amount) || amount <= 0)
    return ctx.reply("Amount must be a positive number.");
  if (!Number.isInteger(amount)) return ctx.reply("Stock quantities must be whole numbers.");

  const info = await resolveSymbol(db, args[1]);
  if (!info)
    return ctx.reply(`Unknown stock ticker "${args[1].toUpperCase()}".\n${comingSoonMessage()}`);

  let price;
  try {
    price = await getStockPrice(info.symbol);
  } catch {
    return ctx.reply("Could not fetch price. Try again later.");
  }
  if (price == null) return ctx.reply(`No price available for ${info.symbol}.`);

  const cost = amount * price;
  const p = await getPortfolio(db, uid(ctx));
  const result = deductCurrency(p, "VND", cost);
  if (!result.ok) {
    return ctx.reply(
      `Insufficient VND. Need ${formatVND(cost)}, have ${formatVND(result.balance)}.`,
    );
  }
  addAsset(p, info.symbol, amount);
  await savePortfolio(db, uid(ctx), p);
  await ctx.reply(
    `Bought ${formatStock(amount)} ${info.symbol} @ ${formatVND(price)}\nCost: ${formatVND(cost)}`,
  );
}

/** /trade_sell <amount> <symbol> — sell VN stock back to VND */
export async function handleSell(ctx, db) {
  const args = parseArgs(ctx);
  if (args.length < 2)
    return usageReply(ctx, "/trade_sell <qty> <TICKER>\nExample: /trade_sell 100 TCB");
  const amount = Number(args[0]);
  if (!Number.isFinite(amount) || amount <= 0)
    return ctx.reply("Amount must be a positive number.");
  if (!Number.isInteger(amount)) return ctx.reply("Stock quantities must be whole numbers.");

  const symbol = args[1].toUpperCase();
  const p = await getPortfolio(db, uid(ctx));
  const result = deductAsset(p, symbol, amount);
  if (!result.ok) return ctx.reply(`Insufficient ${symbol}. You have: ${formatStock(result.held)}`);

  let price;
  try {
    price = await getStockPrice(symbol);
  } catch {
    return ctx.reply("Could not fetch price. Try again later.");
  }
  if (price == null) return ctx.reply(`No price available for ${symbol}.`);

  const revenue = amount * price;
  addCurrency(p, "VND", revenue);
  await savePortfolio(db, uid(ctx), p);
  await ctx.reply(
    `Sold ${formatStock(amount)} ${symbol} @ ${formatVND(price)}\nRevenue: ${formatVND(revenue)}`,
  );
}

/** /trade_convert — disabled, coming soon */
export async function handleConvert(ctx) {
  await ctx.reply(`Currency exchange is not available yet.\n${comingSoonMessage()}`);
}

File: internal/modules/gold/handlers.go
package gold

import (
	"context"
	"errors"
	"strconv"
	"strings"

	"github.com/go-telegram/bot"
	"github.com/go-telegram/bot/models"

	"github.com/tiennm99/miti99bot/internal/log"
	"github.com/tiennm99/miti99bot/internal/modules/util/chathelper"
)

var (
	errInsufficientVND  = errors.New("gold: insufficient VND")
	errInsufficientGold = errors.New("gold: insufficient gold")
)

func (s *state) handlePrice(ctx context.Context, b *bot.Bot, update *models.Update) error {
	args := argsAfterCommand(update.Message.Text)
	if len(args) != 0 {
		return chathelper.Reply(ctx, b, update.Message, "Usage: /gold_price")
	}
	// Fetch under a reply-reserved sub-context (the composite fetcher may try
	// providers sequentially); reply on the original ctx so delivery keeps its
	// budget headroom.
	fetchCtx, cancel := chathelper.FetchContext(ctx)
	defer cancel()
	p, err := s.prices.FetchPrice(fetchCtx)
	if err != nil {
		return s.replyPriceError(ctx, b, update, err)
	}
	lines := goldPriceLines(p)
	return chathelper.Reply(ctx, b, update.Message, strings.Join(lines, "\n"))
}

func goldPriceLines(p GoldPrice) []string {
	return []string{
		"Gold Spot Price (SJC)",
		"Buy: " + FormatVND(p.SJC.Buy) + "/luong",
		"Sell: " + FormatVND(p.SJC.Sell) + "/luong",
	}
}

func (s *state) handleTopup(ctx context.Context, b *bot.Bot, update *models.Update) error {
	userID, ok := senderInfo(update)
	if !ok {
		return chathelper.Reply(ctx, b, update.Message,
			"Cannot identify user - gold only works in private/group chats with a sender.")
	}
	args := argsAfterCommand(update.Message.Text)
	if len(args) != 1 {
		return chathelper.Reply(ctx, b, update.Message, "Usage: /gold_topup <amount>\nExample: /gold_topup 5000000")
	}
	amount, ok := parsePositiveFinite(args[0])
	if !ok || !isSafeVND(amount) {
		return chathelper.Reply(ctx, b, update.Message, "Amount must be a positive finite number within the supported range.")
	}

	defer s.locks.Acquire(strconv.FormatInt(userID, 10))()
	p, err := UpdatePortfolio(ctx, s.store, userID, s.now().UnixMilli(), func(p *Portfolio) error {
		p.AddVND(amount)
		p.Meta.Invested += amount
		return nil
	})
	if err != nil {
		log.Error("gold_save_portfolio", "user", userID, "err", err)
		return chathelper.Reply(ctx, b, update.Message, "Could not save gold portfolio. Try again later.")
	}
	return chathelper.Reply(ctx, b, update.Message,
		"Topped up "+FormatVND(amount)+".\nBalance: "+FormatVND(p.VND))
}

func (s *state) handleBuy(ctx context.Context, b *bot.Bot, update *models.Update) error {
	userID, ok := senderInfo(update)
	if !ok {
		return chathelper.Reply(ctx, b, update.Message,
			"Cannot identify user - gold only works in private/group chats with a sender.")
	}
	args := argsAfterCommand(update.Message.Text)
	if len(args) != 1 {
		return chathelper.Reply(ctx, b, update.Message, "Usage: /gold_buy <luong>\nExample: /gold_buy 1")
	}
	qty, ok := parsePositiveFinite(args[0])
	if !ok {
		return chathelper.Reply(ctx, b, update.Message, "Luong must be a positive finite number.")
	}
	_, sellPrice, err := s.prices.FetchLuongPrices(ctx)
	if err != nil {
		return s.replyPriceError(ctx, b, update, err)
	}
	cost := qty * sellPrice
	if !isSafeVND(cost) {
		return chathelper.Reply(ctx, b, update.Message, "Trade value is too large.")
	}

	defer s.locks.Acquire(strconv.FormatInt(userID, 10))()
	var insufficientBalance *float64
	p, err := UpdatePortfolio(ctx, s.store, userID, s.now().UnixMilli(), func(p *Portfolio) error {
		ok, balance := p.DeductVND(cost)
		if !ok {
			insufficientBalance = &balance
			return errInsufficientVND
		}
		p.AddLuong(qty)
		return nil
	})
	if errors.Is(err, errInsufficientVND) && insufficientBalance != nil {
		return chathelper.Reply(ctx, b, update.Message,
			"Insufficient VND. Need "+FormatVND(cost)+", have "+FormatVND(*insufficientBalance)+".")
	}
	if err != nil {
		log.Error("gold_save_portfolio", "user", userID, "err", err)
		return chathelper.Reply(ctx, b, update.Message, "Could not save gold portfolio. Try again later.")
	}
	return chathelper.Reply(ctx, b, update.Message,
		"Bought "+FormatLuong(qty)+" luong gold @ "+FormatVND(sellPrice)+"/luong\nCost: "+FormatVND(cost)+
			"\nRemaining: "+FormatVND(p.VND))
}

func (s *state) handleSell(ctx context.Context, b *bot.Bot, update *models.Update) error {
	userID, ok := senderInfo(update)
	if !ok {
		return chathelper.Reply(ctx, b, update.Message,
			"Cannot identify user - gold only works in private/group chats with a sender.")
	}
	args := argsAfterCommand(update.Message.Text)
	if len(args) != 1 {
		return chathelper.Reply(ctx, b, update.Message, "Usage: /gold_sell <luong>\nExample: /gold_sell 0.5")
	}
	qty, ok := parsePositiveFinite(args[0])
	if !ok {
		return chathelper.Reply(ctx, b, update.Message, "Luong must be a positive finite number.")
	}
	buyPrice, _, err := s.prices.FetchLuongPrices(ctx)
	if err != nil {
		return s.replyPriceError(ctx, b, update, err)
	}

	defer s.locks.Acquire(strconv.FormatInt(userID, 10))()
	revenue := qty * buyPrice
	if !isSafeVND(revenue) {
		return chathelper.Reply(ctx, b, update.Message, "Trade value is too large.")
	}
	var insufficientHeld *float64
	p, err := UpdatePortfolio(ctx, s.store, userID, s.now().UnixMilli(), func(p *Portfolio) error {
		ok, held := p.DeductLuong(qty)
		if !ok {
			insufficientHeld = &held
			return errInsufficientGold
		}
		p.AddVND(revenue)
		return nil
	})
	if errors.Is(err, errInsufficientGold) && insufficientHeld != nil {
		return chathelper.Reply(ctx, b, update.Message,
			"Insufficient gold. You have: "+FormatLuong(*insufficientHeld)+" luong")
	}
	if err != nil {
		log.Error("gold_save_portfolio", "user", userID, "err", err)
		return chathelper.Reply(ctx, b, update.Message, "Could not save gold portfolio. Try again later.")
	}
	return chathelper.Reply(ctx, b, update.Message,
		"Sold "+FormatLuong(qty)+" luong gold @ "+FormatVND(buyPrice)+"/luong\nRevenue: "+FormatVND(revenue)+
			"\nRemaining: "+FormatVND(p.VND))
}

func (s *state) handleStats(ctx context.Context, b *bot.Bot, update *models.Update) error {
	userID, ok := senderInfo(update)
	if !ok {
		return chathelper.Reply(ctx, b, update.Message,
			"Cannot identify user - /gold_stats needs a sender.")
	}
	p, err := LoadPortfolio(ctx, s.store, userID, s.now().UnixMilli())
	if err != nil {
		log.Error("gold_load_portfolio", "user", userID, "err", err)
		return chathelper.Reply(ctx, b, update.Message, "Could not load gold portfolio. Try again later.")
	}

	lines := []string{"Gold Account Summary\n", "VND: " + FormatVND(p.VND), "Gold: " + FormatLuong(p.Luong) + " luong"}
	totalValue := p.VND
	// Fetch under a reply-reserved sub-context; reply on the original ctx.
	fetchCtx, cancel := chathelper.FetchContext(ctx)
	defer cancel()
	if buyPrice, _, err := s.prices.FetchLuongPrices(fetchCtx); err == nil {
		goldValue := p.Luong * buyPrice
		totalValue += goldValue
		lines = append(lines, "Price: "+FormatVND(buyPrice)+"/luong")
		lines = append(lines, "Gold value: "+FormatVND(goldValue))
		lines = append(lines, "Total value: "+FormatVND(totalValue))
		lines = append(lines, "Invested: "+FormatVND(p.Meta.Invested))
		lines = append(lines, "P&L: "+FormatPnL(totalValue, p.Meta.Invested))
	} else {
		lines = append(lines, "Price: no price")
		lines = append(lines, "Total value: "+FormatVND(totalValue)+" + gold holdings")
	}
	return chathelper.Reply(ctx, b, update.Message, strings.Join(lines, "\n"))
}
